`signMultisigParticipant` & `verifyWithPublicKey` functions added into wallet api

naughtyfox opened the pull request:
`wallet2::sign` function can't work in multisignature wallets

naughtyfox commented:
@moneromooo-monero what do you mean? i've checked it.

moneromooo-monero commented:
Your change uses a single key, which is obviously wrong for a multisignature wallet, no ? Am I missing somehting ?

naughtyfox commented:
In case of multisig we can't use public key because we don't have full private spend. But what we do can is to use signer key + private partial spend (which is cryptographic dependent).
In multisig wallet the only way to identify the wallet participant is signer key. Hence we can sign message with signer public key + private partial spend and verify only by signer public key.
What i do miss is to make verify function that accepts signer key to verify signature.
Please let me know if you agree with my arguments to make me work further on this feature.

moneromooo-monero commented:
That's pointless, since any one participant will be able to sign something, which defeats the point. Wait for luigi to come up with the right way. I asked long ago, but he's always busy so might take some more time.

naughtyfox commented:
> since any one participant will be able to sign something, which defeats the point

I got what you mean. I assume we sign message with certain multisig wallet participant's personal signature (signatures from different participants are not equal) and you think a message is signed from entire wallet perspective. 
We need to define what the wallet signature is - personal participant's or entire wallet.

stoffu commented:
What use case do you have in mind? Why is this needed?

TBH I've been finding the existing sign/verify feature not particularly useful either; it doesn't seem much more useful than popular signature tools like GPG. I'm happy to be educated if I'm missing some important use cases.

moneromooo-monero commented:
I bvelieve onishin wanted to use this to allow miners to set persistent config (such as payout threshold). This depends on just the public identity being the address, and no further information is needed. Instructions can then be checked to have been signed with the address' controller.

moneromooo-monero commented:
If you need this (signing by any one participant), then it can be added as another command, which is multisig specific.

stoffu commented:
Could you provide a little more detail about the multisig procedure you’re developing? I still don’t see where and why participants would want/need to sign arbitrary messages (other than the exchanged key info).

naughtyfox commented:
The service provides ability to exchange keys and partial images between multisig participants. And the service needs to identify certain participant somehow. Obviously the most natural way to do this is to use public signer key (that's what i made previous PR #3599 for). And in order to build authentication system upon signer public key we have to prove the sender has corresponding private key - that's why we need this kind of signature.

naughtyfox commented:
yes, and we want to sign this way our custom messages in service communicating protocol. For example:
```
{
 “destination_address”: “44AFFq5kSiGBoZ4NMDwYtN18obc8AemS33DBLWs3H7otXft3XjrpDtQGv7SqSsaBYBb98uNbr2VBBEt7f2wfn3RVGQBEP3A”,
 “description”: “monero team donation”,
 “signed_transaction”: “xxxxxxxxxxx”,
 “proposal_id”: 1312,
 “amount”: 1.2352,
 “fee”: 0.378
}
```

moneromooo-monero commented:
Are you OK with a multisig specific call for this ? I think using the existing sign functionality is dangerous since it would create an expectation that the threshold was met (and there will probably be some code to do this later).

moneromooo-monero commented:
That is not what I said though.

stoffu commented:
I still don’t understand the need for signing arbitrary messages by multisig participants, because the fact that a multisig tx is partially signed by the participants already means that the tx is genuinely approved by the participants.

Can you explain what kind of forgery is possible without this ability?

naughtyfox commented:
We want to authenticate our users with this signature (+ public signer key) and be sure the message remains unchanged. A message may not contain signed by default data such as transaction or multisig info, it may contain only service data like "synchronization status" or "i agree to sign transaction proposal".

For example, a participant sends message "my synchronization status x/y", adds his signer public key to HTTP header and signs the message with his private key. This way the service will it's participant Alice and the message hasn't been changed by third parties.

PS Sorry for late answer

stoffu commented:
Thanks for explaining, I'm now OK with adding such a signature tool. However, as it stands now, the change in the `sign` code will confuse the `verify` code because the public key against which this signature is to be checked is no longer the wallet public address, but the multisig signer's public key. I.e. one can no longer verify the signature in simplewallet as:

    verify <filename> <address> <signature

As such, some more coding is necessary, e.g. introducing a new variant of `sign/verify` specifically for multisig.

moneromooo-monero commented:
tbh I'd rather you'd keep to the underscore type naming for the wallet2 part (for the GUI it uses that kind of naming).

naughtyfox commented:
@stoffu @moneromooo-monero Hey guys! I have updated this pull request so please check this out.

When i was implementing verify signature function i decided that it'd be better if it was more common and called it `verifyMessageWithPublicKey` because we pass a public key either way.
I introduced new magic for this signature - `SigPkV1` because despite signature algorithm remains the same we use different keys and i'd like to let client know if he's using wrong verify function.
I checked it works (i had to merge the branch with my previous pull request locally).

moneromooo-monero commented on src/wallet/wallet2.cpp:
```diff
@@ -115,6 +115,8 @@ using namespace cryptonote;
 #define STAGENET_SEGREGATION_FORK_HEIGHT 1000000
 #define SEGREGATION_FORK_VICINITY 1500 /* blocks */
 
+static const std::string PUB_KEY_SIGNATURE_MAGIC = "SigPkV1";
```
How about making this have multisig in the name (or some shortening thereof) for clarity ?

naughtyfox replied:
I thought about it, but verification function accepts public key so it potentially maybe a common function and i didn't want to bind it to mulitisig only

naughtyfox replied:
Or maybe you mean variable name (not the magic itself)? If so probably it worth to do it...

moneromooo-monero replied:
I meant the magic. This is multisig specific, and a multisig wallet may be able to sign with proper threshold in the future.

naughtyfox replied:
Okay, will do it in a day

moneromooo-monero commented on src/wallet/wallet2.cpp:
```diff
@@ -9149,6 +9151,40 @@ bool wallet2::verify(const std::string &data, const cryptonote::account_public_a
+bool wallet2::verify_with_public_key(const std::string &data, const crypto::public_key &public_key, const std::string &signature) const
+{
+  if (signature.size() < PUB_KEY_SIGNATURE_MAGIC.size() || signature.substr(0, PUB_KEY_SIGNATURE_MAGIC.size()) != PUB_KEY_SIGNATURE_MAGIC) {
+    LOG_PRINT_L0("Signature header check error");
```
I'd rather new logs use the severity based api (MERROR, MWARNING, etc) as L0 just asumes warning since it can't tell.

naughtyfox replied:
valuable notice, will fix it

naughtyfox commented:
@moneromooo-monero check this out please, i fixed signature magic
